radix-engine/src/system/kernel_modules/transaction_limits/module.rs

```diff
@@ -82,20 +122,34 @@ impl KernelModule for TransactionLimitsModule {
         // push new empty wasm memory value refencing current call frame to internal stack
         api.kernel_get_module_state()
             .transaction_limits
-            .wasm_memory_usage_stack
-            .push(0);
+            .call_frames_stack
+            .push(CallFrameLimitInfo::default());
         Ok(())
     }
 
     fn after_pop_frame<Y: KernelModuleApi<RuntimeError>>(api: &mut Y) -> Result<(), RuntimeError> {
         // pop from internal stack
         api.kernel_get_module_state()
             .transaction_limits
-            .wasm_memory_usage_stack
+            .call_frames_stack
             .pop();
         Ok(())
     }
 
+    fn on_read_substate<Y: KernelModuleApi<RuntimeError>>(
+        api: &mut Y,
+        _lock_handle: LockHandle,
+        _size: usize,
+    ) -> Result<(), RuntimeError> {
+        let tlimit = &mut api.kernel_get_module_state().transaction_limits;
+
+        // Increase read coutner.
+        tlimit.substate_store_read += 1;
+
+        // Validate
+        tlimit.validate_substates()
+    }
```

Comment:
Do we want two limits? one for READ and one for WRITE?

Reply:
I though of separate limits for read and write. Write implemented in similar way as read - using `on_write_substate` event.

Reply:
Haha, didn't read the PR title carefully. It's about substate read only in this PR.